### ISSUE TITLE
MPI+opencl detect only one gpu but fork+opencl detect all gpus on the same node

### ISSUE THREAD

**Issue by quanmqiu** [openwall/john#4659]
I compiled 1.9.0-jumbo-1 on a cluster with CentOS 7.9 and Intel oneAPI compiler.

There are two GPUs in a node.

If I run
```bash
mpirun -np 2 ../run/john --format=oldoffice-opencl --devices=gpu hash.txt
```
I got
```bash
Device 1@g0021: Tesla V100-SXM2-16GB
Device 1@g0021: Tesla V100-SXM2-16GB
Node numbers 1-2 of 2 (MPI)
```

But if I run
```bash
../run/john -fork=2 --format=oldoffice-opencl --devices=gpu hash.txt
```
I got
```bash
Node numbers 1-2 of 2 (fork)
Device 1@g0021: Tesla V100-SXM2-16GB
Device 2@g0021: Tesla V100-SXM2-16GB
```

When I use MPI, only one gpu was detected. But if I use fork, both gpus  were detected.

How can I use MPI + OpenCL on a hpc cluster which nodes have multi-gpu devices?

I tried
```bash
mpirun -machinefile machinefile.txt ../run/john --mask='?l?l?l?l' --format=oldoffice-opencl --devices=gpu hash.txt
```
But only one gpu was detected and used on each node.

Any suggestion?

Thanks.

### System configuration ###
- `$ ./john --list=build-info`.
```bash
Version: 1.9.0-jumbo-1
Build: linux-gnu 64-bit x86_64 SSE2 AC MPI + OMP
SIMD: SSE2, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
$JOHN is ../run/
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
SINGLE_IDX_MAX: 2147483648
SINGLE_BUF_MAX: 4294967295
Effective limit: Number of salts vs. SingleMaxBufferSize
Max. Markov mode level: 400
Max. Markov mode password length: 30
icc version: 20.2.1 (gcc 9.3.1 compatibility)
GNU libc version: 2.17 (loaded: 2.17)
OpenCL headers version: 1.2
Crypto library: OpenSSL
OpenSSL library version: 0100020bf
OpenSSL 1.0.2k  26 Jan 2017     (loaded: OpenSSL 1.0.2k-fips  26 Jan 2017)
GMP library version: 6.0.0
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's
```

- `$ ./john --list=opencl-devices`.
```bash
Platform #0 name: NVIDIA CUDA, version: OpenCL 1.2 CUDA 10.2.141
    Device #0 (1) name:     Tesla V100-SXM2-16GB
    Device vendor:          NVIDIA Corporation
    Device type:            GPU (LE)
    Device version:         OpenCL 1.2 CUDA
    Driver version:         440.64.00 [recommended]
    Native vector widths:   char 1, short 1, int 1, long 1
    Preferred vector width: char 1, short 1, int 1, long 1
    Global Memory:          16160 MB (ECC)
    Global Memory Cache:    2560 KB
    Local Memory:           48 KB (Local)
    Constant Buffer size:   64 KB
    Max memory alloc. size: 4040 MB
    Max clock (MHz):        1530
    Profiling timer res.:   1000 ns
    Max Work Group Size:    1024
    Parallel compute cores: 80
    CUDA cores:             5120  (80 x 64)
    Speed index:            7833600
    Warp size:              32
    Max. GPRs/work-group:   65536
    Kernel exec. timeout:   no
    NVML id:                0
    PCI device topology:    05:00.0
    PCI lanes:              16/16
    Temperature:            77°C
    Utilization:            91%

    Device #1 (2) name:     Tesla V100-SXM2-16GB
    Device vendor:          NVIDIA Corporation
    Device type:            GPU (LE)
    Device version:         OpenCL 1.2 CUDA
    Driver version:         440.64.00 [recommended]
    Native vector widths:   char 1, short 1, int 1, long 1
    Preferred vector width: char 1, short 1, int 1, long 1
    Global Memory:          16160 MB (ECC)
    Global Memory Cache:    2560 KB
    Local Memory:           48 KB (Local)
    Constant Buffer size:   64 KB
    Max memory alloc. size: 4040 MB
    Max clock (MHz):        1530
    Profiling timer res.:   1000 ns
    Max Work Group Size:    1024
    Parallel compute cores: 80
    CUDA cores:             5120  (80 x 64)
    Speed index:            7833600
    Warp size:              32
    Max. GPRs/work-group:   65536
    Kernel exec. timeout:   no
    NVML id:                1
    PCI device topology:    84:00.0
    PCI lanes:              16/16
    Temperature:            35°C
    Utilization:            0%
```

**Comment by solardiz** [comment]
Unrelated to your question, but you really ought to use the latest from this repo and not the 1.9.0-jumbo-1 release, for at least two reasons:

1. You report this on GitHub, and for GitHub issues we assume they're reported against our latest code, not against a release. In fact, our new issue template explicitly says so. You do say yours is against the release, which at least avoids the confusion, but we're nevertheless mostly interested in issues pertaining to the latest. (And our use of GitHub issues is literally to track issues with the program for us to fix, not to provide user support. We use the john-users mailing list for the latter.)

2. There was a fatal known bug affecting oldoffice-opencl in the release, which we've since fixed. `doc/NEWS` here says:
 
```
Major changes from 1.9.0-jumbo-1 (May 2019) in this bleeding-edge version:

- Bug fixes, most importantly in OldOffice-opencl format, which could give
  false negatives on some or many builds.  [Solar, magnum, others; 2019]
```

As to your actual question, I guess @magnumripper will comment on it and determine whether there's anything for us to improve in JtR in this respect.

**Comment by ghost** [comment]
I would say `README.mpi` has the answer:
```
    The following is rejected - you can't use -fork and mpirun [with -np > 1]
    at the same time:
        mpirun -np 4 ./john -fork=...

    This is somewhat more advanced, it will start 1-4/4 on one remote node:
        mpirun -host hostname -np 1 ./john -fork=4 ...

    This will do the same as above, but somewhat less efficiently:
        mpirun -host hostname -np 4 ./john ...
```

So, (me thinks) you should use fork:
```
mpirun -machinefile machinefile.txt '-np 1' ../run/john '-fork=2' --mask='?l?l?l?l' --format=oldoffice-opencl --devices=gpu hash.txt '-node=1-2/4'
```
****
Anyway, I would say you found a misbehavior. The command below is "Poor man's multi-device support" using MPI. 

`gpu_id` should be 1 and 2 (internal values 0 and 1). Could you check the output of `nvidia_smi` when running?
* `mpirun -np 2 ../run/john --format=oldoffice-opencl --devices=gpu hash.txt`
****
IMO, in fact, considering the `but somewhat less efficiently` note, I would prefer `--fork`.

**Comment by ghost** [comment]
Unrelated note:
* recently, someone asked me for support for cracking zip. I advised him to update to bleeding (also important to ZIP format), but he ignored my requests.

Maybe people using a version they shouldn't is happening more often than we would like.

**Comment by magnumripper** [comment]
> I compiled 1.9.0-jumbo-1 on a cluster with CentOS 7.9 and Intel oneAPI compiler.
> 
> There are two GPUs in a node.
> 
> If I run
> ```bash
> mpirun -np 2 ../run/john --format=oldoffice-opencl --devices=gpu hash.txt
> ```
> I got
> ```bash
> Device 1@g0021: Tesla V100-SXM2-16GB
> Device 1@g0021: Tesla V100-SXM2-16GB
> Node numbers 1-2 of 2 (MPI)
> ```
> 
> But if I run
> ```bash
> ../run/john -fork=2 --format=oldoffice-opencl --devices=gpu hash.txt
> ```
> I got
> ```bash
> Node numbers 1-2 of 2 (fork)
> Device 1@g0021: Tesla V100-SXM2-16GB
> Device 2@g0021: Tesla V100-SXM2-16GB
> ```

That's weird. `mpirun -np 2` *should* use both devices exactly like `--fork=2` does (both using eg. `--dev=gpu`). Start out by using latest bleeding-jumbo and we'll go from there.

I haven't ever seen anything like this, but various MPI environments do behave somewhat differently and/or have different defaults - it could be that you need to throw in some extra option to mpirun.

**Comment by quanmqiu** [comment]
Thanks all.

I have changed to use the lastest git bleeding-jumbo.

As magnum said. I did dozens of tests by using different combinations of compilers and mpi implements in last few days.

MPICH + gcc 4/5/6/7/8/9 : failed
MPICH + Intel Compiler 2017/2018/2019/2021 : failed
Intel MPI( based on mpich ) in Parallel Studio 2017/2018/2019/2021 : failed
OpenMPI + gcc 4/5/6/7/8/9 : successed
OpenMPI + Intel Compiler 2017/2018/2019/2021 :  successed

So, it seems that mpich is the source of this problem.

It is not jtr's issue.

After rebuild JtR with openmpi4 + gcc9, a mpi parallel job has run successfully crossing 16 nodes, 2 GPU per node, a total of 32 nvidia V100 GPU. While running, the GPU Util of each V100 was 75% on average.

BTW, the result is wrong.

hash.txt
```bash
$oldoffice$0*027b5bd40f31c876861fe641bf603306*61fac97e2fcdd8e45f3437365eacb13e*9f09d3adc5b190ea7256d50a0cf8d9b4
```
command
```bash
mpirun -np 32 ../run/john --mask='?l?l?l?l?l?l?l?l?l?l?l?l' --format=oldoffice-opencl --devices=gpu hash.txt
```

On a single node, JtR's result of oldoffice-opencl format is always `asfrazsdoaaa`.

If crossing nodes mpi, JtR's result of oldoffice-opencl format is always  `pibglyeohjwn`.

HashCat gived the result correctly with -m 9700.

**Comment by solardiz** [comment]
> BTW, the result is wrong.

Have you upgraded to our latest code from this repo? Like I wrote in here before, it is known and expected that oldoffice-opencl is broken in the 1.9.0-jumbo-1 release.

**Comment by solardiz** [comment]
Also, in what way is the result wrong? Is `asfrazsdoaaa` correct or wrong? Is `pibglyeohjwn` correct or wrong? Are possibly both correct (isn't this the 40-bit old Office stuff, with collisions practically relevant)? What do you expect as the correct result?

The 75% GPU utilization is something to look into, too. But first of all, we need a confirmation that all of your results are now for our latest code, not for the release.

**Comment by quanmqiu** [comment]
@solardiz Version: 1.9.0-jumbo-1+bleeding-d5c7c57

`asfrazsdoaaa` and `pibglyeohjwn` are both wrong, can't decrypt the .xls file.

The correct password is `nnacqpbalaaa` given by hashcat-6.1.1.

**Comment by ghost** [comment]
The format produces collsions (https://github.com/openwall/john/issues/4483#issuecomment-731755664). I wonder if other tools have a better check to produce (just) the correct password.

```
$ john hash.txt -word=words.txt --keep-guessing
Note: Will keep guessing even after finding a possible candidate.
Warning: detected hash type "oldoffice", but the string is also recognized as "oldoffice-opencl"
Use the "--format=oldoffice-opencl" option to force loading these as that type instead
Using default input encoding: UTF-8
Loaded 1 password hash (oldoffice, MS Office <= 2003 [MD5/SHA1 RC4 32/64])
Cost 1 (hash type) is 0 for all loaded hashes
Will run 8 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
MITM key: 09d83a1877
asfrazsdoaaa     (?)     
MITM key: 09d83a1877
pibglyeohjwn     (?)     
MITM key: 09d83a1877
nnacqpbalaaa     (?)     
3g 0:00:00:00 DONE (2021-04-13 09:39) 6.666g/s 48.88p/s 48.88c/s 48.88C/s asfrazsdoaaa..hashcat-6.1.1
Session completed.
```

```
$ ../run/john ~/hash.txt -word=$HOME/words.txt --format=opencl --keep
Device 3: Tahiti [AMD Radeon HD 7900 Series]
Note: Will keep guessing even after finding a possible candidate.
Using default input encoding: UTF-8
Loaded 1 password hash (oldoffice-opencl, MS Office <= 2003 [MD5/SHA1 RC4 OpenCL])
Cost 1 (hash type) is 0 for all loaded hashes
Note: This format may be a lot faster with --mask acceleration (see doc/MASK).
LWS=64 GWS=1048576 (16384 blocks) 
Press 'q' or Ctrl-C to abort, almost any other key for status
Warning: Only 22 candidates left, minimum 1048576 needed for performance.
asfrazsdoaaa     (?)     
pibglyeohjwn     (?)     
nnacqpbalaaa     (?)     
3g 0:00:00:01 DONE (2021-04-13 15:45) 1.648g/s 12.08p/s 12.08c/s 12.08C/s asfrazsdoaaa..hashcat-6.1.1
Session completed. 
```

**Comment by solardiz** [comment]
Looks like we need to prioritize the discussion/work in/on #4483, but meanwhile shouldn't we add `FMT_NOT_EXACT` to the oldoffice formats?

And yes, we need to check how hashcat possibly avoids producing false positives here.

**Comment by solardiz** [comment]
> we need to check how hashcat possibly avoids producing false positives here.

I could be wrong, but at first glance it looks like hashcat only has the full MD5 check, whereas we also have a meet-in-the-middle check, which can result in us detecting a positive earlier (and should probably result in us having better speeds?)

In both programs, and in our case before both checks, there's the 40-bit narrow pipe. So both programs should be affected by collisions. I suppose the question might be then whether all collisions resulting in a reported guess correspond to working passwords (including possibly different ones from what the user had set) or only some of them do. Maybe our MITM check somehow allows wrong passwords to pass, whereas without it only correctly-colliding passwords would pass?

Can/should we possibly enhance our MITM check not to produce a final result, but merely to proceed to the slow path, @magnumripper? Wouldn't that bring the best of both worlds (higher speeds yet no false positives - only actually usable colliding passwords)?

**Comment by magnumripper** [comment]
> MPICH + gcc 4/5/6/7/8/9 : failed
> MPICH + Intel Compiler 2017/2018/2019/2021 : failed
> Intel MPI( based on mpich ) in Parallel Studio 2017/2018/2019/2021 : failed
> OpenMPI + gcc 4/5/6/7/8/9 : successed
> OpenMPI + Intel Compiler 2017/2018/2019/2021 : successed
> 
> So, it seems that mpich is the source of this problem.

Oh, right. It was caused by c2dbd45 (#1920) and the problem is that there's no way (to my knowledge) to see the "local world size" under MPICH. That is, we don't know how many other nodes, if any, run on the same host.

c2dbd45 was a fix for mscash2-opencl only so I think I can tweak this, I'll look into it.

**Comment by magnumripper** [comment]
> Looks like we need to prioritize the discussion/work in/on #4483, but meanwhile shouldn't we add `FMT_NOT_EXACT` to the oldoffice formats?

Oldoffice has collisions with the property that any of the cracked passwords can be used to open the file (I've seen it and done it many many times) so no, it shouldn't have `FMT_NOT_EXACT`.

> 
> And yes, we need to check how hashcat possibly avoids producing false positives here.

I'm pretty sure there *isn't any way* to verify that a certain positive guess is the actual password that was used when saving the document! This case must be either a bug or user error/confusion of some sort.

> I could be wrong, but at first glance it looks like hashcat only has the full MD5 check, whereas we also have a meet-in-the-middle check, which can result in us detecting a positive earlier (and should probably result in us having better speeds?)

Hashcat has them separated as several different formats: One that doesn't do any MITM, one that only finds the MITM key, one that uses the MITM key to look for alternative passwords.

> In both programs, and in our case before both checks, there's the 40-bit narrow pipe. So both programs should be affected by collisions. I suppose the question might be then whether all collisions resulting in a reported guess correspond to working passwords (including possibly different ones from what the user had set) or only some of them do. Maybe our MITM check somehow allows wrong passwords to pass, whereas without it only correctly-colliding passwords would pass?

That'd be a bug I haven't seen surface before. @quanmqiu could you share the complete office file with us? Or second best, with only me in a private email?

> Can/should we possibly enhance our MITM check not to produce a final result, but merely to proceed to the slow path, @magnumripper? Wouldn't that bring the best of both worlds (higher speeds yet no false positives - only actually usable colliding passwords)?

I'm pretty sure we always do the full check if a candidate passed early rejection. If that's not the case, it's definitely a bug!

**Comment by solardiz** [comment]
BTW, looks like our CPU format doesn't use the MITM in MD5 mode, only using it in SHA-1 mode. Our OpenCL format uses it in both modes. Something to fix in the CPU format?

> I'm pretty sure we always do the full check if a candidate passed early rejection.

We have things like this:

```opencl
        if (cs->has_mitm) {
                *result = (key[0] == cs->mitm[0] && key[1] == cs->mitm[1]);
                if (*result)
                        atomic_xchg(&cs->cracked, 1);
        } else {
```

So the MITM-based result, whatever it is, becomes final - not just early rejection.

> If that's not the case, it's definitely a bug!

I'm not so sure: if the MITM value is necessarily correct, then the computation in the `else` branch is guaranteed to have produced the same result, no? And if there are cases when the MITM value is not correct, then it's not usable for early rejection either.

Anyway, according to Claudio's comment above there's no difference in results from the CPU and OpenCL formats here, and since the CPU format doesn't appear to use MITM (it only updates the MITM field, but never depends on what's in there), the issue reported here (if any) must be different.

**Comment by magnumripper** [comment]
> BTW, looks like our CPU format doesn't use the MITM in MD5 mode, only using it in SHA-1 mode. Our OpenCL format uses it in both modes. Something to fix in the CPU format?

I just saw that too, except I can't see SHA-1 do any early-rejection either? It only reports the value. However in SHA-1 mode there's an **optional** extra field that can be used to tell a collision from the real password.

**Comment by solardiz** [comment]
> I can't see SHA-1 do any early-rejection either?

In the SHA-1 case, `cmp_all` starts by reading from the `mitm` field before proceeding with further computation:

```c
                        RC4_set_key(&key, 16, mitm_key[index]); /* dek */
```

Not the same as early rejection, though.

**Comment by magnumripper** [comment]
I just verified that hashcat too finds all three passwords. Their `--keep-guessing` isn't working though, so you need to test them in three separate runs (it will just report the first found).

**Comment by magnumripper** [comment]
> However in SHA-1 mode there's an **optional** extra field that can be used to tell a collision from the real password.

I wonder if this extra field feature can be present for a MD5 document as well? That would explain why only one password actually works. We'd need a sample such document then, for tweaking oldoffice2john and the formats.

**Comment by solardiz** [comment]
Added "bug" label for c2dbd45 affecting formats other than mscash2-opencl.

**Comment by solardiz** [comment]
> I wonder if this extra field feature can be present for a MD5 document as well? That would explain why only one password actually works.

Makes sense. BTW, with salt type 3, which is SHA-1 with 40-bit intermediate key, when the extra field is present this means we may sometimes find a working encryption key, but end up not reporting it because the password presumably wouldn't pass. Maybe in such cases we should be reporting the encryption key anyway (but keep guessing passwords), so that if one has software capable of key input (I think some proprietary password recovery and forensics tools have that, maybe some free software ones, too?) the user could use the key to decrypt the document without ever cracking the right password.

**Comment by quanmqiu** [comment]
I'm sorry. The origin office excel file is a private file of my friend. Two weeks before, he asked me help for password recovery but didn't want to give me the source file. I had to teach him to use office2john.py. Then he send me just a hash.txt.

First I got the correct password by using HashCat on a single node with 4 * V100. I hope to speed up computation time by using more GPUs, but HashCat can't run crossing nodes. So I tried to use JtR + mpi.

Today I run the version 1.9.0-jumbo-1+bleeding-85978db for hours with option --keep-guessing, got a lot of collisions:
```bash
pibglyeohjwn -
aueysylnhlof
iisqfjbthlku
wcdogibhahlo
eaabneumehjw
lcqxdaodujmx
tcenjohpujmx
fnfjhybxbjmd
tnmgzzgoklke
mytzntogulof
asfrazsdoaaa -
gxdekzlpoaaa
kfgtuerwoahj
drmuzjpsnahl
khigcfbdihlo
rkmzhjjxkjmd
xlftdzsijlot
uqdpriqglhjm
ziagdghxglke
foimxbvrnhjw
spwmmmgovjwg
rpdjyidrshlk
qhkatuvwjlot
bwovhspysaah
xznbzmgiglku
qlxceoixnhjw
uhjgeubsglku
nrhtjrffplke
nnacqpbalaaa --
zliqyuevvlot
bxgxqfekjlke
bellwipumhlk
tmpaimewmaah
lsjbmxsgtahl
iamcnnhhfjmd
chafxwfjfjmd
rjkpyjytshlo
lcovlqcnmahj
vexbipypshlo
zzfqrxnvdaah
......
```

**Comment by magnumripper** [comment]
So was he using the latest `office2john.py` from here, or the Jumbo-1 release version? I *think* in this case it should not matter but to be sure, if he was using the older one, you should make him generate a new hash with the current version and see if they end up identical.

**Comment by quanmqiu** [comment]
> 
> 
> So was he using the latest `office2john.py` from here, or the Jumbo-1 release version? I _think_ in this case it should not matter but to be sure, if he was using the older one, you should make him generate a new hash with the current version and see if they end up identical.

Yes, the hash.txt was generated by the lastest `office2john.py` downloaded from github web.

**Comment by magnumripper** [comment]
I did some testing with `Office/Office_Secrets/Office_XP/XP-Office97-2000_myhovercraftisfullofeels_.xls` from https://github.com/openwall/john-samples. That file is also an excel file of type 0 so should be identical to the one @quanmqiu attacks. The correct password for it is `myhovercraftisfullofeels_` but that old version truncated the password to length 15, so would crack as `myhovercraftisf`.

The MITM hash is `bcca60ccc9` and the first collision I found (using `-mask:?l -len=12`) was `sfiioswjaaaa`. And that password indeed opens the sample document.

**Comment by solardiz** [comment]
> First I got the correct password by using HashCat on a single node with 4 * V100. I hope to speed up computation time by using more GPUs, but HashCat can't run crossing nodes. So I tried to use JtR + mpi.

If you already "got the correct password" and were doing this as password recovery for a friend, then why did you still want to speed up the computation? Did this "correct password" not work?

> Today I run the version 1.9.0-jumbo-1+bleeding-85978db for hours with option --keep-guessing, got a lot of collisions:

Our expectation is that all/any of these passwords should actually work. What is the actual behavior your friend is seeing - do any of them work, and if so which ones?

**Comment by quanmqiu** [comment]
> 
> 
> If you already "got the correct password" and were doing this as password recovery for a friend, then why did you still want to speed up the computation? Did this "correct password" not work?

The `nnacqpbalaaa` did decrypted his origin excel file. 

Just because he was a lucky boy. HashCat predicted that it needs more than 309 days to complete full search in the 12 low case letters space by using 4*V100. But in fact I got `nnacqpbalaaa` after 11 minutes.

That's why I want to get a solution of MPI + GPU. 

> Our expectation is that all/any of these passwords should actually work. What is the actual behavior your friend is seeing - do any of them work, and if so which ones?

I will get more collisions send to him for verifying.

**Comment by solardiz** [comment]
> Just because he was a lucky boy.

More likely because this is not the original password, but is one of many collisions, and those do indeed work for decrypting the file.

**Comment by quanmqiu** [comment]
> 
> 
> > Just because he was a lucky boy.
> 
> More likely because this is not the original password, but is one of many collisions, and those do indeed work for decrypting the file.

I'm using hashcat with --keep-guessing to search fully. It may need some times before I can report a usable result.

**Comment by solardiz** [comment]
As magnum wrote in a comment above:

> I just verified that hashcat too finds all three passwords. Their --keep-guessing isn't working though, so you need to test them in three separate runs (it will just report the first found).

I don't know whether it's a bug (edit: I mean their --keep-guessing not working) in some specific version(s) of hashcat only, or in all of them, or something specific to magnum's testing. If it's a bug in current hashcat, we probably need to report it to them, and ideally even send them a pull request with a fix, @magnumripper?

**Comment by magnumripper** [comment]
> As magnum wrote in a comment above:
> 
> > I just verified that hashcat too finds all three passwords. Their --keep-guessing isn't working though, so you need to test them in three separate runs (it will just report the first found).
> 
> I don't know whether it's a bug (edit: I mean their --keep-guessing not working) in some specific version(s) of hashcat only, or in all of them, or something specific to magnum's testing. If it's a bug in current hashcat, we probably need to report it to them, and ideally even send them a pull request with a fix, @magnumripper?

I didn't research it further but I assume it happened because all collisions were in the same batch of work (one textfile with all positives). But yes I'll report it to them.

**Comment by magnumripper** [comment]
I was correct, known problem and wontfix: https://github.com/hashcat/hashcat/issues/1879

**Comment by solardiz** [comment]
> I was correct, known problem and wontfix: [hashcat/hashcat#1879](https://github.com/hashcat/hashcat/issues/1879)

OK, although the piece of hashcat documentation quoted there says "This does not mean that valid passwords are skipped; they are always reported", which per your description here and the description there I think is wrong.

**Comment by magnumripper** [comment]
True, but that must be down to weird wording in that document.

The OP said exactly what I see happen: "It returns only one plain-text in one kernel loop".

That document currently says (apart from your quote) "This limitation cannot be fixed, because it would require too much device (GPU/CPU) memory. If we wanted to report back all possible password candidates executed in a single kernel invocation, it would require this much memory: (...)".

**Comment by magnumripper** [comment]
@quanmqiu you might want to upgrade to latest code (3088ea62d) for some GPU boost.

**Comment by quanmqiu** [comment]
> 
> 
> @quanmqiu you might want to upgrade to latest code ([3088ea6](https://github.com/openwall/john/commit/3088ea62da65d4033f7f49c56fe6264328dc166c)) for some GPU boost.

Updated to 3088ea6, build with openmpi 4.0.5 + gcc9.3.

```bash
mpirun -np 16 ../run/john --mask='?l?l?l?l?l?l?l?l?l?l?l?l' --format=oldoffice-opencl --devices=gpu hash.txt --keep-guessing
```

Total 16 * V100 on 8 nodes, after 6 hours running, GPU-Util of each card can keep >95% on average.

```bash
Thu Apr 15 17:31:12 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.64.00    Driver Version: 440.64.00    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:05:00.0 Off |                    0 |
| N/A   62C    P0   207W / 300W |    743MiB / 16160MiB |     98%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  On   | 00000000:06:00.0 Off |                    0 |
| N/A   64C    P0   219W / 300W |    743MiB / 16160MiB |     97%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     13590      C   ../run/john                                  423MiB |
|    0     13591      C   ../run/john                                  309MiB |
|    1     13590      C   ../run/john                                  309MiB |
|    1     13591      C   ../run/john                                  423MiB |
+-----------------------------------------------------------------------------+
```
